data/transactions/logic/eval.go

```diff
@@ -559,9 +572,20 @@ func EvalContract(program []byte, gi int, aid basics.AppIndex, params *EvalParam
 		Txn:          &params.TxnGroup[gi],
 		appID:        aid,
 	}
+
+	if cx.Proto.IsolateClearState && cx.Txn.Txn.OnCompletion == transactions.ClearStateOC {
+		if cx.PooledApplicationBudget != nil && *cx.PooledApplicationBudget < cx.Proto.MaxAppProgramCost {
+			return false, nil, &ClearStateBudgetError{*cx.PooledApplicationBudget}
+		}
+	}
+
 	pass, err := eval(program, &cx)
 
-	// update side effects
+	// update side effects. It is tempting, and maybe even a good idea, to store
+	// the pointer to cx.scratch instead.  Since we don't modify them again,
+	// it's probably safe. However it may have poor GC characteristics (because
+	// we'd be storing a pointer into a much larger structre, the cx), and
```

Comment:
@jannotti Minor - Looks like a typo.

```suggestion
	// we'd be storing a pointer into a much larger structure, the cx), and
```